cirq/ops/pauli_string.py

```diff
@@ -22,43 +23,85 @@
 class PauliString(raw_types.Operation):
 
-    def __init__(self,
-                 qubit_pauli_map: Optional[
-                     Mapping[raw_types.Qid, pauli_gates.Pauli]] = None,
-                 coefficient: Union[int, float, complex] = 1) -> None:
-        if qubit_pauli_map is None:
-            qubit_pauli_map = {}
-        qubit_pauli_map = {
-            q: p
-            for q, p in qubit_pauli_map.items()
-            if not isinstance(p, common_gates.IdentityGate)
-        }
-        for p in qubit_pauli_map.values():
-            if not isinstance(p, pauli_gates.Pauli):
-                raise TypeError(f'{p} is not a Pauli')
-        self._qubit_pauli_map = dict(qubit_pauli_map)
-        self._coefficient = complex(coefficient)
+    def __init__(
+            self,
+            *contents: PAULI_STRING_LIKE,
+            qubit_pauli_map: Optional[Dict['cirq.Qid', 'cirq.Pauli']] = None,
+            coefficient: Union[int, float, complex] = 1):
+        """Initializes a new PauliString.
+
+        Args:
+            *contents: A value or values to convert into a pauli string. This
+                can be a number, a pauli operation, a dictionary from qubit to
+                pauli/identity gates, or collections thereof. If a list of
+                values is given, they are each individually converted and then
+                multiplied from left to right in order.
+            qubit_pauli_map: Initial dictionary mapping qubits to pauli
+                operations. Defaults to the empty dictionary. Note that, unlike
+                dictionaries passed to contents, this dictionary must not
+                contain any identity gate values. Further note that this
+                argument specifies values that are logically *before* factors
+                specified in `contents`; `contents` are *right* multiplied onto
+                the values in this dictionary.
+            coefficient: Initial scalar coefficient. Defaults to 1.
+
+        Examples:
+            >>> a, b, c = cirq.LineQubit.range(3)
+
+            >>> print(cirq.PauliString([cirq.X(a), cirq.X(a)]))
+            I
+
+            >>> print(cirq.PauliString(-1, cirq.X(a), cirq.Y(b), cirq.Z(c)))
+            -X(0)*Y(1)*Z(2)
+
+            >>> print(cirq.PauliString({a: cirq.X}, [-2, 3, cirq.Y(a)]))
+            -6j*Z(0)
+
+            >>> print(cirq.PauliString({a: cirq.I, b: cirq.X}))
+            X(1)
+
+            >>> print(cirq.PauliString({a: cirq.Y},
+            ...                        qubit_pauli_map={a: cirq.X}))
+            1j*Z(0)
+        """
+        if qubit_pauli_map is not None:
+            for v in qubit_pauli_map.values():
+                if not isinstance(v, pauli_gates.Pauli):
+                    raise TypeError(f'{v} is not a Pauli')
+
+        p = _MutablePauliString(coef=complex(coefficient),
+                                paulis=dict(qubit_pauli_map or {}))
+        p.inline_times_pauli_string_like(contents)
```

Comment:
seems weird to use both `qubit_pauli_map` and `contents`. You can imagine wanting to deprecate / remove `qubit_pauli_map` completely in the future and having everything put in `contents`. These semantics would make that change very painful

Reply:
There are two reasons for leaving it present.

1. It is slightly faster and more predictable because it involves less type inspection shenanigans.
2. If you remove it, you are quite a large risk of creating an infinite recursion loop (e.g. if you accidentally create an X operation in the initializer, you'll infinitely recurse due to the call to `PauliString.__init__` inside `SingleQubitPauliStringGateOperation.__init__`).

Reply:
I disagree that these changes make later deprecation change more painful. We can deprecate the argument in the same way we'd deprecate any other, with an auto-rewrite that puts it at the start of the contents list.

Reply:
The use of the constructor for implicit multiplying is a consequence of the idea "I should be able to pass in a list of Pauli operations and it understands what I mean" + closure.